crates/biome_css_parser/src/syntax/mod.rs

```diff
@@ -0,0 +1,138 @@
+use crate::parser::CssParser;
+use biome_css_syntax::CssSyntaxKind::*;
+use biome_css_syntax::{CssSyntaxKind, T};
+use biome_parser::diagnostic::expected_any;
+use biome_parser::parse_recovery::ParseRecovery;
+use biome_parser::prelude::ParsedSyntax::{Absent, Present};
+use biome_parser::prelude::{ParseDiagnostic, ParsedSyntax, ToDiagnostic};
+use biome_parser::{token_set, Parser, ParserProgress, TokenSet};
+use biome_rowan::TextRange;
+
+const SELECTOR_RECOVERY_SET: TokenSet<CssSyntaxKind> = token_set![T!['{'], T!['}'],];
+const BODY_RECOVERY_SET: TokenSet<CssSyntaxKind> =
+    SELECTOR_RECOVERY_SET.union(token_set![T![.], T![*],]);
+
+pub(crate) fn parse_root(p: &mut CssParser) {
+    let m = p.start();
+
+    parse_rules_list(p).expect("Parse rule list, handle this case");
+
+    m.complete(p, CSS_ROOT);
+}
+
+pub(crate) fn parse_rules_list(p: &mut CssParser) -> ParsedSyntax {
+    let rules = p.start();
+    while !p.at(EOF) {
+        match p.cur() {
+            T![.] => {
+                parse_rule(p).expect("Parse rule, handle this case properly");
+            }
+            _ => return Absent,
+        }
+    }
+
+    let completed = rules.complete(p, CSS_RULE_LIST);
+
+    Present(completed)
+}
+
+pub(crate) fn parse_rule(p: &mut CssParser) -> ParsedSyntax {
+    let m = p.start();
+    if parse_selector_list(p)
+        .or_recover(
+            p,
+            &ParseRecovery::new(CSS_BOGUS_PATTERN, SELECTOR_RECOVERY_SET),
+            expect_pattern,
+        )
+        .is_err()
+    {
+        m.abandon(p);
```

Comment:
Could you please share your strategy for situations where we might need to abandon and return as absent, rather than completing and returning bogus kind?

Reply:
I don't have a **clear** strategy in mind yet, but I've had some ideas while starting this work, still it's best if we discuss them.

For rules, I thought we could bogus+recover in the prelude and in the block.

This would allow us to parse cases like:

```css
.foo }
.bar {}
```

```
.686 {}
.bar {}
```

Although, I'm not sure about the prelude of the rule.

Or maybe we should have a big `CssBogusRule`? What do you think?

Reply:
I agree with you, let's try to use a recovery strategy.